### PR TITLE
eskip: reuse lexer and parser

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
```
name               old time/op    new time/op    delta
ParsePredicates-8    4.80µs ± 4%    3.11µs ± 8%  -35.27%  (p=0.000 n=10+9)

name               old alloc/op   new alloc/op   delta
ParsePredicates-8    6.22kB ± 0%    0.74kB ± 0%  -88.03%  (p=0.000 n=10+10)

name               old allocs/op  new allocs/op  delta
ParsePredicates-8      37.0 ± 0%      35.0 ± 0%   -5.41%  (p=0.000 n=10+10)
```

### CODE REVIEW

File: eskip/eskip.go [hunk 1]
```diff
@@ -591,11 +592,28 @@ func newRouteDefinition(r *parsedRoute) (*Route, error) {
 	return rd, err
 }
 
+type eskipLexParser struct {
+	lexer  eskipLex
+	parser eskipParserImpl
+}
```
Comment (AlexanderYastrebov) [inline eskip/eskip.go:598]: Parser consists of 1+16 `eskipSymType` structs
https://github.com/zalando/skipper/blob/77f4c5e868917a90a8823d734e7186fb7a34140d/eskip/parser.go#L200-L204
which is quite large
https://github.com/zalando/skipper/blob/77f4c5e868917a90a8823d734e7186fb7a34140d/eskip/parser.go#L21-L44

A new lexer is always used with the parser so it makes sense to pool them together.

`eskipSymType` is in fact a union type, i.e. only one of the fields is used on each parse step.
I've experimented with implementing it as a 
```yacc
%union {
	val interface{}
}
```
and a set of converters like
```go
func (u *eskipSymType) token() string        { return u.val.(string) }
func (u *eskipSymType) stringvals() []string { return u.val.([]string) }
func (u *eskipSymType) route() *parsedRoute { return u.val.(*parsedRoute) }
...
```
similar to how it is done in [cockroachdb](https://github.com/cockroachdb/cockroach/blob/c21c90f93219b857858518d25a8bc061444d573c/pkg/sql/parser/sql.y#L205-L249).

This brings some improvement but requires quite heavy rewrite of parser.y grammar rules.

I have also couple of lexer refactorings that eliminate inefficiencies and reduce allocations but not as much as pooling lexer and parser. These could be done in subsequent PRs.

File: eskip/eskip.go [hunk 2]
```diff
@@ -591,11 +592,28 @@ func newRouteDefinition(r *parsedRoute) (*Route, error) {
...
 // executes the parser.
 func parse(code string) ([]*parsedRoute, error) {
-	l := newLexer(code)
-	eskipParse(l)
-	return l.routes, l.err
+	lp, ok := parserPool.Get().(*eskipLexParser)
+	if !ok {
+		lp = &eskipLexParser{}
+	} else {
+		*lp = eskipLexParser{}
```
Comment (szuecs) [inline eskip/eskip.go:608]: Didn't we get the lp already? Why do we need to "reset" the value?

**Comment by szuecs** [comment]
:+1:

**Comment by AlexanderYastrebov** [comment]
:+1:

**Pull request merged**